workflow-service/src/main/java/com/redhat/parodos/project/service/ProjectServiceImpl.java

```diff
@@ -118,4 +125,52 @@ public List<ProjectResponseDTO> getProjectsByUsername(String username) {
 				.map(projectUserRole -> modelMapper.map(projectUserRole, ProjectResponseDTO.class)).toList();
 	}
 
+	@Transactional
+	public ProjectUserRoleResponseDTO updateUserRolesToProject(UUID id, List<UserRoleRequestDTO> userRoleRequestDTOs) {
+		Project project = projectRepository.findById(id)
+				.orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "project not found"));
+
+		Set<ProjectUserRole> projectUserRoles = userRoleRequestDTOs.stream()
+				.map(userRoleRequestDTO -> Map.entry(
+						userService.getUserEntityByUsername(userRoleRequestDTO.getUsername()),
+						roleRepository.findByNameIgnoreCase(userRoleRequestDTO.getRole().name())))
+				.filter(userRoleMap -> userRoleMap.getKey() != null && userRoleMap.getValue().isPresent())
+				.map(userRoleMap -> ProjectUserRole.builder()
+						.id(ProjectUserRole.Id.builder().projectId(project.getId()).userId(userRoleMap.getKey().getId())
+								.roleId(userRoleMap.getValue().get().getId()).build())
+						.project(project).user(userRoleMap.getKey()).role(userRoleMap.getValue().get()).build())
+				.collect(Collectors.toSet());
+
+		projectUserRoleRepository.deleteAllByIdProjectIdAndIdUserIdIn(project.getId(),
+				projectUserRoles.stream().map(projectUserRole -> projectUserRole.getUser().getId()).toList());
+		project.setProjectUserRoles(projectUserRoles);
+		projectUserRoleRepository.saveAll(projectUserRoles);
```

Comment:
Can you explain the delete & save here?

Reply:
the mapping between projectUserrole and project is not bidrectional. what I updated in project/projectUserrole will be not reflected to the other in db. so I have to delete and refill. this is the known limition of embedded collection